File: src/java1/algorithms/strings/reverseWordsInString/ReverseWordsInString.java
package java1.algorithms.strings.reverseWordsInString;

public class ReverseWordsInString {

    //Two pointer array reversal TC: O(n) SC: O(n)
    private static String reverseWordsInString1(String str) {
        String[] words = str.trim().split("\\s+");
        int i = 0, j = words.length - 1;

        while (i < j) {
            String temp = words[i];
            words[i] = words[j];
            words[j] = temp;
            i++;
            j--;
        }
        return String.join(" ", words);
    }

    //String builder array reversal TC: O(n) SC: O(n)
    private static String reverseWordsInString2(String str) {
        String[] words = str.trim().split("\\s+");

        StringBuilder sb = new StringBuilder();
        for (int i = words.length - 1; i >= 0; i--) {
            sb.append(words[i]);
            if(i != 0){
                sb.append(" ");               
            }
        }

        return sb.toString();
    }

    public static void main(String[] args) {
        String str1 = "It is fun to learn DSA";
        String str2 = "  hello DSA  ";
        System.out.println(reverseWordsInString1(str1));
        System.out.println(reverseWordsInString1(str2));

        System.out.println(reverseWordsInString2(str1));
        System.out.println(reverseWordsInString2(str2));
    }
}

File: src/javascript/algorithms/strings/reverseWordsInString/reverseWordsInString.js
//TC: O(n) SC: O(n)
function reverseWordsInString1(str) {
  let words = str.trim().split("/s+/");
  let result = "";

  for (let i = words.length - 1; i >= 0; i--) {
    result += words[i] + " ";
  }

  return result.slice(0, result.length - 1);
}

//TC: O(n) SC: O(1)
function reverseWordsInString2(str) {
  let words = str.trim().split("/\s+/");
  let i = 0, j = str.length - 1;

  while (i <= j) {
    [words[i], words[j]] = [words[j], words[i]];
    i++;
    j--;
  }

  return words.join("");
}

let str1 = "the sky is blue";
let str2 = "  hello world  ";
console.log(reverseWordsInString1(str1));
console.log(reverseWordsInString1(str2));

console.log(reverseWordsInString2(str1));
console.log(reverseWordsInString2(str2));
